Add Caps lock message to login screen

nikunj8866 opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/48345

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

Commit:
Add caps lock warning on login screen

nikunj8866 commented:
> The following accounts have interacted with this PR and/or linked issues. I will continue to update these lists as activity occurs. You can also manually ask me to refresh this list by adding the `props-bot` label.
> 
> ## Unlinked Accounts
> The following contributors have not linked their GitHub and WordPress.org accounts: @[nhatkar@updraftplus.com](mailto:nhatkar@updraftplus.com).
> 
> Contributors, please [read how to link your accounts](https://make.wordpress.org/core/2020/03/19/associating-github-accounts-with-wordpress-org-profiles/) to ensure your work is properly credited in WordPress releases.
> 
> **To understand the WordPress project's expectations around crediting contributors, please [review the Contributor Attribution page in the Core Handbook](https://make.wordpress.org/core/handbook/best-practices/contributor-attribution-props/).**

Please use @nikunj8866 WordPress profile.

nikunj8866 commented:
### After

![image](https://github.com/user-attachments/assets/c7eaaf24-e03f-44c2-b696-3ca6e121ce6f)

Commit:
Fix javaScript coding standards

nikunj8866 committed:
Added caps warning for Add/Edit Users password field

nikunj8866 committed:
Changed warning color and bypass warning in macOS

nikunj8866 committed:
Changed notification color to yellow, Added screen readers capabilities and used KeyboardEvent.getModifierState() for checking

SirLouen commented on src/js/_enqueues/admin/user-profile.js:
```diff
@@ -203,6 +229,30 @@
+			 * Shows a warning if Caps Lock is detected. Also uses wp.a11y.speak()
+			 * for screen readers. Hides the warning on blur or when Caps Lock is not on.
+			 */
+			var $capsWarning = $( '#caps-warning' );
```
This section is completely identical to the one in L92. 
Can't we unify both into a function to avoid repetition of the same code?

nikunj8866 committed:
Optimize JS code to remove identical code

nikunj8866 committed:
Remove mac chrome from the skip

nikunj8866 committed:
update js doc comment

joedolson commented on src/js/_enqueues/admin/user-profile.js:
```diff
@@ -322,6 +328,53 @@
+		// Skip warning on macOS Safari (they show native indicators).
+		if (
+			isMac &&
+			( ua.indexOf( 'safari' ) !== -1 )
```
Firefox on Mac will now show this, as well. So probably need another exception.

And Chrome does not show the Caps Lock icon, but has 'Safari' in its User Agent string, so this check is catching Chrome.

<img width="369" height="264" alt="Screenshot 2025-09-05 at 2 02 45 PM" src="https://github.com/user-attachments/assets/495fadff-201e-47ea-b90a-1d25a6077377" />

joedolson commented on src/wp-admin/css/forms.css:
```diff
@@ -719,6 +719,61 @@ fieldset label,
+	z-index: 10;
+	box-shadow: 0 1px 3px rgba(0,0,0,0.1);
+}
+
```
Stylistically, I'd like to simplify this. I don't think it needs to be floating, first - so let's give it `position: relative`.

Along with that, I'd remove the left positioning and make the top positioning -8px just to erase the password input's bottom margin and remove the box shadow.

The z-index would now be irrelevant, and the font-size can be inherited. Text color should be `3c434a` to match other text on the same page.

Then we should adjust the colors to use something more in the WordPress color palette. The closest yellow to the current color would be `fcf9e8`, but I'd probably choose `f5e6ab`, instead.

nikunj8866 replied:
@joedolson Thanks for the feedback! I've simplified the styles as suggested. Can you confirm if this design looks correct, or if you'd like further tweaks?

<img width="793" height="177" alt="image" src="https://github.com/user-attachments/assets/82c9f969-cd07-4082-a2b7-8deb36835ee6" />

- In the image below, the icon color has updated:
<img width="913" height="316" alt="image" src="https://github.com/user-attachments/assets/898544c8-c67b-42fb-831e-ec86c2eeb67e" />

joedolson replied:
I prefer the second example; but I'm wondering if we need both the icon and the CSS generated arrow. They don't really look right together, because they aren't aligned - having the arrow points in different horizontal locations looks strange to me. Since we're not trying to follow, I think maybe we should also remove the arrow.

nikunj8866 replied:
**I've removed the CSS-generated arrow.**

<img width="1439" height="954" alt="image" src="https://github.com/user-attachments/assets/dfe3d26f-733f-488b-af1d-a645cbb39a35" />

joedolson commented on src/js/_enqueues/admin/user-profile.js:
```diff
@@ -322,6 +328,53 @@
+		$input.on( 'keydown', function( e ) {
+			if ( isCapsLockOn( e.originalEvent || e ) ) {
+				$capsWarning.show();
+				wp.a11y.speak( __( 'Caps lock is on.' ) );
```
This will fire on every keydown. So if a user *wants* capslock enabled when typing their password, then this will be very irritating. 

The notification needs to be limited, so that it only fires once. 

Additionally, screen readers already announce the state changes between caps lock on and caps lock off, so we should not emit this notification if the caps lock key is pressed; that notification will be handled by the screen reader. Only send the message if a valid password character is typed.

joedolson reviewed (changes requested):
There are some style things I'd like to see changed for consistency with color palette, an accessibility issue, and the browser detection needs some work.

nikunj8866 committed:
fix browser detection issue for macOS and disable checking capslock for every keydown

nikunj8866 committed:
update comment text

joedolson committed:
Merge branch 'trunk' into pr/8726

joedolson commented on src/js/_enqueues/admin/user-profile.js:
```diff
@@ -322,6 +328,53 @@
+				wp.a11y.speak( __( 'Caps lock is on.' ) );
+			} else {
+				$capsWarning.hide();
+			}
```
This is better, but I think that the caps warning should be hidden when capslock is disabled. The spoken message & warning appearance should only happen on character pressing, but if a user observes the warning and turns off capslock, then the message should go away immediately, rather than only after the next character type.

nikunj8866 replied:
Updated the code so that the Caps Lock warning now **hides immediately when Caps Lock is turned off**, even if no new character is typed.

joedolson reviewed (dismissed):
Getting closer!

nikunj8866 committed:
hide warning when capslock off and fix popup design

praful2111 commented:
Tested with the below config 

## Test Report
### Description
This report validates whether the indicated patch works as expected.

Patch tested: (https://github.com/WordPress/wordpress-develop/pull/8726)

### Environment
- WordPress: 6.9-alpha-20250922.141940
- PHP: 7.4.31-dev
- Server: PHP.wasm
- Database: WP_SQLite_Driver (Server: 5.5 / Client: 3.40.1)
- Browser: Chrome 141.0.0.0
- OS: Linux
- Theme: Twenty Twenty-Five 1.3
- MU Plugins: None activated
- Plugins:
  * Test Reports 1.2.0

### Actual Results
1.  ✅ Issue resolved with patch.

### Additional Notes
- Any additional details worth mentioning.

### Supplemental Artifacts
Chrome
<img width="1919" height="933" alt="image" src="https://github.com/user-attachments/assets/36f63037-e7a2-431f-8d5a-9b9eed0f6833" />

Comet
<img width="1855" height="970" alt="image" src="https://github.com/user-attachments/assets/b637dc3d-9f3d-4aa0-916e-7fb969db28bd" />

Firefox
<img width="1887" height="874" alt="image" src="https://github.com/user-attachments/assets/321454ee-0f49-43c0-adaf-922369cf8a25" />

User Profile
<img width="1600" height="789" alt="image" src="https://github.com/user-attachments/assets/ceb72746-4a75-4602-8366-df376c45ef7a" />

joedolson committed:
Merge branch 'trunk' into pr/8726

westonruter commented on src/js/_enqueues/admin/user-profile.js:
```diff
@@ -332,6 +340,75 @@
+			capsLockOn   = false;
+
+		$input.on( 'keydown', function( e ) {
+			var ev = e.originalEvent || e;
```
Let's not use `e` and `ev` abbreviations. Variables should not be abbreviated for readability.

Who knows. Maybe `ev` is about an electric vehicle! 🚗 😄

westonruter replied:
Will `e.originalEvent` ever be empty? I don't believe so. I think this can be simply:

```suggestion
		$input.on( 'keydown', function( jqEvent ) {
			var event = jqEvent.originalEvent;
```

westonruter commented on src/js/_enqueues/admin/user-profile.js:
```diff
@@ -332,6 +340,75 @@
+
+			// Skip CapsLock key itself.
+			if ( ev.key === 'CapsLock' ) {
+				if (capsLockOn) {
```
JS coding standards for whitespace should be followed.

```suggestion
				if ( capsLockOn ) {
```

westonruter commented on src/js/_enqueues/admin/user-profile.js:
```diff
@@ -332,6 +340,75 @@
+		// Skip warning on macOS Safari + Firefox (they show native indicators).
+		if ( isMac && ( isSafari || isFirefox ) ) {
+			return false;
+		}
```
Instead of doing this check here, why not skip adding the listener in the first place when on a Mac?

westonruter commented on src/js/_enqueues/admin/user-profile.js:
```diff
@@ -332,6 +340,75 @@
+	function bindCapsLockWarning( $input ) {
+		var $capsWarning = $( '#caps-warning' ),
+			capsLockOn   = false;
+
```
Per below, we can skip adding the `keydown` event when on a Mac to in the first place.

```suggestion

		// Skip warning on macOS Safari + Firefox (they show native indicators).
		if ( isMac && ( isSafari || isFirefox ) ) {
			return;
		}

```

westonruter commented on src/js/_enqueues/admin/user-profile.js:
```diff
@@ -332,6 +340,75 @@
+
+		if ( typeof e.getModifierState === 'function' ) {
+			return e.getModifierState( 'CapsLock' );
+		}
```
This function does not always return a boolean. Here it returns `void`. Should this add:


```suggestion
			return false;
		}
```

westonruter replied:
On second thought, there's no need for the function exists check. This method is available in all supported browsers: https://caniuse.com/mdn-api_mouseevent_getmodifierstate

```suggestion
		return e.getModifierState( 'CapsLock' );
```

westonruter commented on src/js/_enqueues/admin/user-profile.js:
```diff
@@ -332,6 +340,75 @@
+	 *
+	 * @return {boolean} True if Caps Lock is on, false otherwise. 
+	 */
+	function isCapsLockOn( e ) {
```
TODO: Rename `e` to `event`.

joedolson committed:
Incorporate suggestions from @westonruter

I have some additional changes I want to make to this and haven't seen any activity from @nikunj8866 recently.

joedolson committed:
Caps lock warning improvements

Move caps lock warning output to JS to reduce code duplication; fix issue where changes were not registered in capslock key pressed; improve comments; styling changes.

joedolson committed:
Adjust profile page styles to merge better with existing styling.

joedolson committed:
Fix Safari test so it doesn't return true on Chrome

joedolson committed:
Missing var declarations

joedolson committed:
Change function description to be declarative.

joedolson committed:
Update padding on mobile to match password strength

joedolson committed:
Only fire blur within the same document.

joedolson commented:
Screenshots:

<img width="3807" height="1482" alt="user profile page with capslock warning visible" src="https://github.com/user-attachments/assets/5c466935-fab5-4418-a6a4-c853e0f1e219" />

<img width="3800" height="1500" alt="user login with capslock warning visible" src="https://github.com/user-attachments/assets/a3fbe322-a9f9-42c7-970f-c3603f6a4589" />

westonruter commented on src/js/_enqueues/admin/user-profile.js:
```diff
@@ -332,6 +340,79 @@
+
+		$capsWarning = $( '<div id="caps-warning" class="caps-warning"></div>' );
+		$capsIcon    = $( '<span class="caps-icon" aria-hidden="true"><svg viewBox="0 0 24 26" xmlns="http://www.w3.org/2000/svg" fill="#3c434a" stroke="#3c434a" stroke-width="0.5"><path d="M12 5L19 15H16V19H8V15H5L12 5Z"/><rect x="8" y="21" width="8" height="1.5" rx="0.75"/></svg></span>' );
+		$capsText    = $( '<span class="caps-warning-text">' + __( 'Caps lock is on.' ) + '</span>' );
```
This is safer in general as it avoids injection issues with concatenating HTML.

```suggestion
		$capsText    = $( '<span>', { 'class': 'caps-warning-text', text: __( 'Caps lock is on.' ) } );
```

joedolson committed:
Update src/js/_enqueues/admin/user-profile.js

Co-authored-by: Weston Ruter <westonruter@gmail.com>

joedolson commented:
Closed in r60988
